### PR TITLE
Remove 8.13 deprecations

### PR DESCRIPTION
**Pull request by proux01**
Now that 8.14 is branched, we could think about removing things deprecated in 8.13.

There is still an 8.13 deprecation in `plugins/ltac/g_auto.mlg` that I was unable to address and a few others appearing in the manual (c.f. `git grep "deprecated:: 8.13"`)

Overlays (should be merged before merging this PR):
- ~argosy https://github.com/mit-pdos/argosy/pull/3~ (merged)
- ~bbv https://github.com/mit-plv/bbv/pull/36~ (merged)
- ~cross_crypto https://github.com/mit-plv/cross-crypto/pull/23~ (merged)
- ~Mtac2: https://github.com/Mtac2/Mtac2/pull/340~ (merged)
- ~metacoq: https://github.com/MetaCoq/metacoq/pull/582~ (merged)
- ~fiat_parsers https://github.com/mit-plv/fiat/pull/57~ (merged)
- ~chain of git submodules Kami ( https://github.com/mit-plv/kami/pull/25 ) -> bedrock2 ( mit-plv/bedrock2#196 ) -> fiat_crypto ( https://github.com/mit-plv/fiat-crypto/pull/1013 )~ (all merged)

### CODE REVIEW

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
@silene can we add this to 8.14?

**Comment by silene** [comment]
No, numerous users have already updated their developments for 8.14, so I will not merge any pull cause that might cause breakage with respect to the alpha release of 8.14.

**Comment by proux01** [comment]
Anyway, this is specifically targeting 8.15 since we commit to keep deprecations for 2 releases so: deprecation in 8.13 -> removal in 8.15

**Comment by olaure01** [comment]
Concerning `prod_curry` and `prod_uncurry`, the deprecation attribute in `8.13` was put on a `Definition`. Is it visible from the user side? wouldn't an intermediary step with a `Notation` be necessary?

**Comment by proux01** [comment]
@JasonGross @Zimmi48 is there a good reason for the above minimize messages to be duplicated?

**Comment by proux01** [comment]
@JasonGross I'm sorry, I suck at git submodules and I'm unable to do the fiat_parsers overlay, it seems to be enough to replace `Numeral Notation` by `Number Notation` in `etc/coq-script/compatibilitty/Coq__{8_15,master}__Compat.v`

**Comment by SkySkimmer** [comment]
Why is coqbot repeating itself? @JasonGross

**Comment by Zimmi48** [comment]
> @JasonGross @Zimmi48 is there a good reason for the above minimize messages to be duplicated?

You did manually restart the metacoq job several times, right? At this point, coqbot will print minimize messages every time that a pipeline completes, which explains some duplication (we plan on making coqbot less verbose by checking if it's reposting a comment similar to a previous one).

Then there was an issue with the fiat-parsers job being restarted over and over by coqbot. In principle, it only restarts job that failed spuriously. But here, it restarted a job that failed with `fatal: Couldn't find remote ref rm_deprecated_813` and this is probably because the code checking for spurious failure was too lax in this case.

If this kind of situation happens again, besides tagging me (which is always a good idea), you can also check if the GitLab pipeline has a job which is constantly being retried and cancelling it.

**Comment by proux01** [comment]
Thanks for the explanations and advice, I didn't even thought coqbot could restart jobs and just thought the job was taking a very long time without looking further.

File: doc/sphinx/changes.rst
Comment (Alizter) [inline doc/sphinx/changes.rst:868]: Would it be better to have a "deprecated attributes" page so these can still be referenced? That way these can also be removed easily in the future.

Also do we even need to mention them anymore?

cc @Zimmi48 @jfehrle

File: theories/Init/Datatypes.v
Comment (Alizter) [inline theories/Init/Datatypes.v:285]: Whats going on here? Am I right in suspecting that #[deprecated] doesn't work for definitions, so you have to hide it in a notation? This was something that I thought about implementing (#[deprecated] for more defs,lemma etc.). This looks like a possible use case to me.

File: theories/Program/Combinators.v
Comment (Alizter) [inline theories/Program/Combinators.v:54]: Nit: this is the wrong way round to how it was before:
```suggestion
(** [curry] and [uncurry] are each others inverses. *)
```

File: vernac/vernacentries.ml
Comment (Alizter) [inline vernac/vernacentries.ml:749]: Is this comment still relevant? cc @ejgallego

File: doc/sphinx/changes.rst
Comment (proux01) [reply doc/sphinx/changes.rst:868]: > Would it be better to have a "deprecated attributes" page so these can still be referenced? That way these can also be removed easily in the future.

I really don't know, I don't feel relevant enough about the manual to take such a decision.

> Also do we even need to mention them anymore?

This is the changelog, and IMHO it shouldn't be edited (or minimally, but at least nothing should be removed).

File: theories/Init/Datatypes.v
Comment (proux01) [reply theories/Init/Datatypes.v:285]: Correct, that's following the very relevant point made above https://github.com/coq/coq/pull/14819#issuecomment-908633151

File: theories/Program/Combinators.v
Comment (proux01) [reply theories/Program/Combinators.v:54]: IINM, `prod_curry` was replaced by `uncurry` and `prod_uncurry` by `curry`.

File: vernac/vernacentries.ml
Comment (proux01) [reply vernac/vernacentries.ml:749]: Well, that still looks somewhat redundant to me, the deprecation/removal of old boolean attributes doesn't change that point IMHO.

File: theories/Program/Combinators.v
Comment (Alizter) [reply theories/Program/Combinators.v:54]: OK then this seems wrong to me. AFAIK currying is `prod_rect` (or even `sigT_rect` in general). Currying should take a function `A * B -> C` and give a function `A -> B -> C`. Uncurrying should be the inverse.

Comment (proux01) [reply theories/Program/Combinators.v:54]: I agree with you. So the old `prod_curry` was wrongly named and the new `uncurry` is good. Everything seems fine isn't it?

Comment (Alizter) [reply theories/Program/Combinators.v:54]: Sorry I made a mistake! `prod_rect` should take `A -> B -> C` and give `A * B -> C`, so is infact `uncurry`.

OK, so everything it good. It's unfortunately very easy to confuse this, so I give sympathy to the past author!

**Comment by Alizter** [comment]
@coqbot ci minimize

**Comment by Alizter** [comment]
Hmm seems fiat parsers is still using `Numeral` rather than `Number`. I suspect this is to support an older version of coq. @JasonGross What should be done?

**Comment by Alizter** [comment]
So it looks like we need more overlays.

File: theories/Init/Datatypes.v
Comment (Zimmi48) [reply theories/Init/Datatypes.v:285]: See also https://coq.inria.fr/refman/using/libraries/writing.html#deprecating-library-objects-or-tactics

File: doc/sphinx/changes.rst
Comment (Zimmi48) [reply doc/sphinx/changes.rst:868]: We have always done it like this and IMHO this is fine.

**Comment by Zimmi48** [comment]
The removal should be documented in (probably several distinct) changelog entries.

**Comment by proux01** [comment]
> So it looks like we need more overlays:
> 
> * fiat-parsers: replace `Numeral` with `Number`.
> * fiat-crypto-legacy: replace `Grab Existential Variables.` with `Unshelve`.

Indeed, I'm in the process of doing them, but those things use chains of git submodules, so I cannot add overlays directly here.

File: theories/Init/Datatypes.v
Comment (Alizter) [reply theories/Init/Datatypes.v:285]: @Zimmi48 Right, but this only works for notations but there are many cases where you might want to deprecate something other than a notation.

Comment (Zimmi48) [reply theories/Init/Datatypes.v:285]: Sure

**Comment by Alizter** [comment]
@Zimmi48 Can you tell coqbot to stop minimizing?

**Comment by Zimmi48** [comment]
Done. And agreed.

File: theories/Init/Datatypes.v
Comment (herbelin) [reply theories/Init/Datatypes.v:285]: I think that what the reference manual is saying is that this could directly be:
```
#[deprecated(since = "8.13", note = "Use curry instead.")]
Notation prod_uncurry := curry.
```
But it depends also on how important it is for compatibility to preserve the unnaming in `prod_uncurry` of arguments `f`, `x`, `y` of `curry`. Maybe the `_subdef` indirection gives a higher compatibility at the end.

Comment (proux01) [reply theories/Init/Datatypes.v:285]: Indeed, directly using `Notation prod_uncurry := curry.` would cause breakages due to a change in implicit arguments status between (the old) `prod_uncurry` and (the new) `curry`.

File: doc/changelog/03-notations/14819-rm_deprecated_813.rst
Comment (Zimmi48) [inline doc/changelog/03-notations/14819-rm_deprecated_813.rst:2]: Removed is already part of the sentence, no need to repeat it.

File: doc/changelog/03-notations/14819-rm_deprecated_813.rst
Comment (Zimmi48) [inline doc/changelog/03-notations/14819-rm_deprecated_813.rst:10]: Probably worth quoting filenames too.

File: doc/changelog/03-notations/14819-rm_deprecated_813.rst
Comment (Zimmi48) [inline doc/changelog/03-notations/14819-rm_deprecated_813.rst:22]: There's one too many quotes after `notemplate`.

File: doc/changelog/03-notations/14819-rm_deprecated_813.rst
Comment (Zimmi48) [inline doc/changelog/03-notations/14819-rm_deprecated_813.rst:26]: Maybe say "deprecated in 8.13" (because like this it reads as if `Unshelve` had been introduced in 8.13).

File: doc/changelog/03-notations/14819-rm_deprecated_813.rst
Comment (proux01) [reply doc/changelog/03-notations/14819-rm_deprecated_813.rst]: Thanks (I had vague remembrance of a previous discussion a while ago on the topic that ended up to the conclusion that the word should be duplicated).

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by Zimmi48** [comment]
@JasonGross I think this PR was merged before fcl was fixed. I'll disable the comments when fcl fails to avoid spamming every PR.

**Comment by proux01** [comment]
@Alizter this was ready, but ideally we should have waited for all the overlays to be merged before merging the PR itself, since all overlays are backward compatible.

**Comment by Alizter** [comment]
@proux01 Yes, sorry about that, I should have waited. @Zimmi48 already know, and is helping me sort it out.

**Comment by proux01** [comment]
No worries, there were only two overlays left, so this should be easily fixed.

**Comment by JasonGross** [comment]
@Zimmi48 I don't see a fcl overlay, only one for fiat-crypto (non-legacy)...

**Comment by Zimmi48** [comment]
Right, it looks like it was not prepared. Can you merge the fiat-crypto one and see whether you can adapt FCL?

**Comment by proux01** [comment]
@JasonGross indeed, I didn't had time to prepare it before the PR got merged. I guess it amount to replacing `Grab Existential Variables` by `Unshelve` (and reverting the order of the resulting subgoals if there are more than one).

**Comment by JasonGross** [comment]
fcl should be fixed now, so the notifications can be re enabled

**Comment by Zimmi48** [comment]
Done.

**Pull request merged**
